keras_nlp/models/xlm_roberta.py

```diff
@@ -0,0 +1,302 @@
+    """XLM-RoBERTa encoder with a customizable set of hyperparameters.
+
+    This network implements a bi-directional Transformer-based encoder as
+    described in
+    ["Unsupervised Cross-lingual Representation Learning at Scale"](https://arxiv.org/abs/1911.02116).
+    It includes the embedding lookups and transformer layers, but does not
+    include the masked language modeling head used during pretraining.
+
+    This class gives a fully configurable XLM-R model with any number of
+    layers, heads, and embedding dimensions. For specific XLM-R architectures
+    defined in the paper, see, for example, `keras_nlp.models.XLMRoberta`.
+
+    Args:
+        vocabulary_size: int. The size of the token vocabulary.
+        num_layers: int. The number of transformer layers.
+        num_heads: int. The number of attention heads for each transformer.
+            The hidden size must be divisible by the number of attention heads.
+        hidden_dim: int. The size of the transformer encoding layer.
+        intermediate_dim: int. The output dimension of the first Dense layer in
+            a two-layer feedforward network for each transformer.
+        dropout: float. Dropout probability for the Transformer encoder.
+        max_sequence_length: int. The maximum sequence length this encoder can
+            consume. The sequence length of the input must be less than
+            `max_sequence_length`.
+        name: string, optional. Name of the model.
+        trainable: boolean, optional. If the model's variables should be
+            trainable.
+
+    Example usage:
+    ```python
+    # Randomly initialized XLM-R model
+    model = keras_nlp.models.XLMRobertaCustom(
+        vocabulary_size=50265,
+        num_layers=12,
+        num_heads=12,
+        hidden_dim=768,
+        intermediate_dim=3072,
+        max_sequence_length=12
+    )
+
+    # Call encoder on the inputs.
+    input_data = {
+        "token_ids": tf.random.uniform(
+            shape=(1, 12), dtype=tf.int64, maxval=50265),
+        "padding_mask": tf.constant(
+            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)),
+    }
+    output = model(input_data)
+    ```
+    """
+
+    pass
+
+
+PREPROCESSOR_DOCSTRING = """XLM-RoBERTa preprocessor with pretrained vocabularies.
+
+This preprocessing layer will do three things:
+
+ - Tokenize any number of inputs using a
+   `keras_nlp.tokenizers.SentencePieceTokenizer`.
+ - Pack the inputs together using a
+   `keras_nlp.models.roberta.RobertaMultiSegmentPacker` with the appropriate
+   `"<s>"`, `"</s>"` and `"<pad>"` tokens, i.e., adding a single `<s>` at the
+   start of the entire sequence, `[</s>, </s>]` at the end of each segment,
+   save the last and a `</s>` at the end of the entire sequence.
+ - Construct a dictionary of with keys `"token_ids"`, `"padding_mask"`, that can
+   be passed directly to a XLM-RoBERTa model.
+
+The original fairseq implementation implements a very hacky solution to
```

Comment:
I would say this more concisely...

Note the the original fairseq implementation modifies the indices of the sentencepiece tokenizer
output. To preserve vocab compatibility, we make the same change. Keep this in mind if generating
your own vocabulary for tokenization.